File: apps/android/app/src/main/java/com/chatapp/mesh/MeshPower.kt
package com.chatapp.mesh

// MeshPower.kt — battery and resource management for the offline mesh.
//
// Mirrors services/mesh/power.go. Mesh networking is expensive on battery-
// powered devices. This module owns every resource knob in one place so no
// caller ever runs an unrestricted scan loop or an unbounded queue. The
// manager is purely advisory state + pure functions over that state, so
// tests can simulate days of operation deterministically.

/** Device power snapshot as reported by the OS. */
data class PowerState(
    val batteryPercent: Int,  // 0..100, or -1 when unknown
    val charging: Boolean,
    val powerSaver: Boolean,
)

data class ScanIntervals(
    val activeMs: Long = 10_000,
    val idleMs: Long = 30_000,
    val dozeMs: Long = 120_000,
)

data class ConnectionLimits(
    val maxPeers: Int = 8,
    val maxPeersCharging: Int = 16,
    val maxQueuePerLink: Int = 256,
)

class PowerManager(
    private var state: PowerState = PowerState(50, false, false),
    private val intervals: ScanIntervals = ScanIntervals(),
    private val limits: ConnectionLimits = ConnectionLimits(),
) {
    companion object {
        private const val TOPO_ACTIVE_WINDOW_MS = 90_000L
    }

    private val lock = Any()
    private var lastTrafficAt: Long = 0
    private var lastScanAt: Long = 0
    private var topologyAt: Long = 0

    fun setPowerState(s: PowerState) { synchronized(lock) { state = s } }

    fun markTraffic(nowMs: Long) { synchronized(lock) { lastTrafficAt = nowMs } }

    fun markTopologyChange(nowMs: Long) { synchronized(lock) { topologyAt = nowMs } }

    fun nextScanIn(nowMs: Long): Pair<Long, Boolean> = synchronized(lock) {
        val recentTraffic = lastTrafficAt > 0 && nowMs - lastTrafficAt < TOPO_ACTIVE_WINDOW_MS
        val recentTopo = topologyAt > 0 && nowMs - topologyAt < TOPO_ACTIVE_WINDOW_MS
        when {
            state.charging && recentTraffic -> intervals.activeMs to true
            recentTraffic || recentTopo -> intervals.activeMs to true
            state.powerSaver && state.batteryPercent in 0..20 -> intervals.dozeMs to false
            state.batteryPercent in 0..15 -> intervals.dozeMs to false
            else -> intervals.idleMs to false
        }
    }

    fun maxPeers(): Int = synchronized(lock) {
        if (state.charging) limits.maxPeersCharging else limits.maxPeers
    }

    fun maxQueuePerLink(): Int = synchronized(lock) { limits.maxQueuePerLink }

    fun relayScale(): Double = synchronized(lock) {
        when {
            state.charging -> 1.0
            state.powerSaver && state.batteryPercent in 0..20 -> 0.25
            state.batteryPercent in 0..15 -> 0.25
            else -> 1.0
        }
    }

    fun scaleRelayQuota(baseBytes: Long): Long =
        (baseBytes.toDouble() * relayScale()).toLong()

    fun status(): Map<String, Any> = synchronized(lock) {
        mapOf(
            "battery_percent" to state.batteryPercent,
            "charging" to state.charging,
            "power_saver" to state.powerSaver,
            "scan_intervals" to mapOf(
                "active_ms" to intervals.activeMs,
                "idle_ms" to intervals.idleMs,
                "doze_ms" to intervals.dozeMs,
            ),
            "max_peers" to limits.maxPeers,
            "max_peers_chg" to limits.maxPeersCharging,
            "max_queue" to limits.maxQueuePerLink,
            "relay_scale" to relayScale(),
        )
    }
}

File: apps/ios/ChatApp/Mesh/MeshPower.swift
import Foundation

/// MeshPower.swift — battery and resource management for the offline mesh.
///
/// Mirrors services/mesh/power.go. Owns every resource knob — scan cadences,
/// connection cap, relay quota scaling — so no caller runs an unrestricted
/// scan loop. Purely advisory state + pure functions; deterministic in tests.

struct PowerState {
    var batteryPercent: Int = 50
    var charging = false
    var powerSaver = false
}

struct ScanIntervals {
    var activeMs: Int64 = 10_000
    var idleMs: Int64 = 30_000
    var dozeMs: Int64 = 120_000
}

struct ConnectionLimits {
    var maxPeers = 8
    var maxPeersCharging = 16
    var maxQueuePerLink = 256
}

final class PowerManager {
    private static let topoActiveWindowMs: Int64 = 90_000

    private let lock = NSLock()
    private var state = PowerState()
    private let intervals = ScanIntervals()
    private let limits = ConnectionLimits()
    private var lastTrafficAt: Int64 = 0
    private var lastScanAt: Int64 = 0
    private var topologyAt: Int64 = 0

    func setPowerState(_ s: PowerState) {
        lock.lock(); defer { lock.unlock() }
        state = s
    }

    func markTraffic(nowMs: Int64) {
        lock.lock(); defer { lock.unlock() }
        lastTrafficAt = nowMs
    }

    func markTopologyChange(nowMs: Int64) {
        lock.lock(); defer { lock.unlock() }
        topologyAt = nowMs
    }

    func nextScanIn(nowMs: Int64) -> (intervalMs: Int64, fullScan: Bool) {
        lock.lock(); defer { lock.unlock() }
        let recentTraffic = lastTrafficAt > 0 && nowMs - lastTrafficAt < Self.topoActiveWindowMs
        let recentTopo = topologyAt > 0 && nowMs - topologyAt < Self.topoActiveWindowMs
        switch true {
        case state.charging && recentTraffic: return (intervals.activeMs, true)
        case recentTraffic || recentTopo: return (intervals.activeMs, true)
        case state.powerSaver && state.batteryPercent <= 20: return (intervals.dozeMs, false)
        case state.batteryPercent <= 15: return (intervals.dozeMs, false)
        default: return (intervals.idleMs, false)
        }
    }

    func maxPeers() -> Int {
        lock.lock(); defer { lock.unlock() }
        return state.charging ? limits.maxPeersCharging : limits.maxPeers
    }

    func maxQueuePerLink() -> Int {
        lock.lock(); defer { lock.unlock() }
        return limits.maxQueuePerLink
    }

    func relayScale() -> Double {
        lock.lock(); defer { lock.unlock() }
        switch true {
        case state.charging: return 1.0
        case state.powerSaver && state.batteryPercent <= 20: return 0.25
        case state.batteryPercent <= 15: return 0.25
        default: return 1.0
        }
    }

    func scaleRelayQuota(baseBytes: Int64) -> Int64 {
        Int64(Double(baseBytes) * relayScale())
    }

    func status() -> [String: Any] {
        lock.lock(); defer { lock.unlock() }
        return [
            "battery_percent": state.batteryPercent,
            "charging": state.charging,
            "power_saver": state.powerSaver,
            "scan_intervals": [
                "active_ms": intervals.activeMs,
                "idle_ms": intervals.idleMs,
                "doze_ms": intervals.dozeMs,
            ],
            "max_peers": limits.maxPeers,
            "max_peers_chg": limits.maxPeersCharging,
            "max_queue": limits.maxQueuePerLink,
            "relay_scale": relayScale(),
        ]
    }
}